xiilib/_gunicorn/charm.py

```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The base Gunicorn charm class for all WSGI application charms."""
+import abc
+import logging
+
+import ops
+from charms.data_platform_libs.v0.data_interfaces import DatabaseRequiresEvent
+from charms.traefik_k8s.v2.ingress import IngressPerAppRequirer
+from pydantic import BaseModel  # pylint: disable=no-name-in-module
+
+from xiilib._gunicorn.charm_state import CharmState
+from xiilib._gunicorn.observability import Observability
+from xiilib._gunicorn.secret_storage import GunicornSecretStorage
+from xiilib._gunicorn.webserver import GunicornWebserver
+from xiilib._gunicorn.wsgi_app import WsgiApp
+from xiilib.database_migration import DatabaseMigration, DatabaseMigrationStatus
+from xiilib.databases import Databases, make_database_requirers
+from xiilib.exceptions import CharmConfigInvalidError
+
+logger = logging.getLogger(__name__)
+
+
+class GunicornBase(abc.ABC, ops.CharmBase):  # pylint: disable=too-many-instance-attributes
+    """Gunicorn-based charm service mixin."""
+
+    @abc.abstractmethod
+    def get_wsgi_config(self) -> BaseModel:
+        """Return the framework related configurations."""
+
+    @abc.abstractmethod
+    def get_cos_dir(self) -> str:
+        """Return the directory with COS related files."""
+
+    def __init__(self, framework: ops.Framework, wsgi_framework: str) -> None:
+        """Initialize the instance.
+
+        Args:
+            framework: operator framework.
+            wsgi_framework: WSGI framework name.
+        """
+        super().__init__(framework)
+        self._secret_storage = GunicornSecretStorage(
+            charm=self, key=f"{wsgi_framework}_secret_key"
+        )
+        self._database_requirers = make_database_requirers(self, self.app.name)
+        try:
+            wsgi_config = self.get_wsgi_config()
+        except CharmConfigInvalidError as exc:
+            self._update_app_and_unit_status(ops.BlockedStatus(exc.msg))
+            return
```

Comment:
We should be careful with return in the constructors as some events can't be triggered (like pebble_ready)

I think we might want to introduce the "collect_status" even handler for this charm (could be done later)

Reply:
If we didn't return in the constructor, then the "charm state" will contain invalid data causing problems in the following execution. This is one of problems of charm state.

Reply:
The reason I approved this is because we're changing the architecture guidance at the moment to address this issue. That hasn't been finished yet so it seems like it should be ok to follow the previous pattern for the moment. We'll need to refactor this once we have settled on the new guidance